### PR TITLE
Closes #580 - Modify dropwizard metrics timers and histograms to use …

### LINKED ISSUE #580: Modify dropwizard metrics timers to use HdrHistogram backed reservoir
For timers and histograms, the default reservoir is the ExponentiallyDecayingReservoir which keeps reporting the same numbers if new data points are not generated.  From the research below, it looks the HdrHistogram is a better solution:

https://lab.getbase.com/aggregating-response-time/
http://www.slideshare.net/tkowalcz/everybody-lies-48319373
https://bitbucket.org/marshallpierce/hdrhistogram-metrics-reservoir/

### PR DESCRIPTION
**Pull request by mikewalch**
…HdrHistogram backed reservoir

 * Fixed config to use directory instead of file for csv reporter
 * Fluo now prevents multiple reporters if HdrHistogramResetOnSnapshotReservoir is used
 * Made reservoir class configurable

### CODE REVIEW

File: modules/core/src/main/java/io/fluo/core/metrics/starters/GraphiteReporterStarter.java [hunk 1]
```diff
@@ -48,6 +52,8 @@
             .convertRatesTo(rateUnit).prefixedWith(prefix).build(graphite);
     reporter.start(config.getInt("frequency", 60), TimeUnit.SECONDS);
 
+    log.info("Reporting metrics to graphite");
```
Comment (keith-turner) [inline modules/core/src/main/java/io/fluo/core/metrics/starters/GraphiteReporterStarter.java]: would be useful to print host and port where metrics are being reported

File: modules/core/src/main/java/io/fluo/core/metrics/starters/CsvReporterStarter.java [hunk 2]
```diff
@@ -42,8 +46,11 @@
 
     CsvReporter reporter =
         CsvReporter.forRegistry(params.getMetricRegistry()).convertDurationsTo(durationUnit)
-            .convertRatesTo(rateUnit).build(new File(file));
+            .convertRatesTo(rateUnit).build(new File(dir));
     reporter.start(config.getInt("frequency", 60), TimeUnit.SECONDS);
+
+    log.info("Reporting metrics as csv");
```
Comment (keith-turner) [inline modules/core/src/main/java/io/fluo/core/metrics/starters/CsvReporterStarter.java]: could add dir to log message

**Comment by keith-turner** [comment]
Need to update metrics.md.  In that docs is mentions that JMX will be setup if no reporter is configured.  Could remove that.

File: modules/core/pom.xml [hunk 3]
```diff
@@ -83,6 +83,14 @@
       <artifactId>zookeeper</artifactId>
     </dependency>
     <dependency>
+      <groupId>org.hdrhistogram</groupId>
```
Comment (keith-turner) [inline modules/core/pom.xml]: what are the license of these?  Does not matter now... but if we ever try to incubate at Apache then Fluo can only depend on certain licenses.

File: modules/core/src/main/java/io/fluo/core/metrics/MetricsUtil.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    } catch (InstantiationException | IllegalAccessException | ClassNotFoundException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  public static Timer getTimer(FluoConfiguration config, MetricRegistry registry, String name) {
+    Timer timer = registry.getTimers().get(name);
+    if (timer == null) {
+      timer = new Timer(getConfiguredReservoir(config));
+      registry.register(name, timer);
```
Comment (keith-turner) [inline modules/core/src/main/java/io/fluo/core/metrics/MetricsUtil.java]: there is a concurrency issue here... two threads could call this and both try to register... I looked at the register code, seems to throw an exception if name is already registered

File: modules/core/src/main/java/io/fluo/core/metrics/starters/GraphiteReporterStarter.java (hunk 1 above)
Comment (mikewalch) [inline modules/core/src/main/java/io/fluo/core/metrics/starters/GraphiteReporterStarter.java]: fixed in e838120692b

File: modules/core/src/main/java/io/fluo/core/metrics/starters/CsvReporterStarter.java (hunk 2 above)
Comment (mikewalch) [inline modules/core/src/main/java/io/fluo/core/metrics/starters/CsvReporterStarter.java]: fixed in e838120692b

File: modules/core/src/main/java/io/fluo/core/metrics/MetricsUtil.java (hunk 4 above)
Comment (mikewalch) [inline modules/core/src/main/java/io/fluo/core/metrics/MetricsUtil.java]: it should be fixed in e838120692b

File: modules/core/pom.xml (hunk 3 above)
Comment (mikewalch) [inline modules/core/pom.xml]: HdrHistogram is BSD and hdrhistogram-metrics-reservoir is Apache 2.0 so it sounds like it should not be an issue.

**Comment by mikewalch** [comment]
Updated JMX mention in metrics.md in e8381206

**Comment by keith-turner** [comment]
+1

**Pull request merged**